api/agent/lb_agent.go

```diff
@@ -183,6 +211,48 @@ func (a *lbAgent) Submit(callI Call) error {
 	return a.handleCallEnd(ctx, call, err, true)
 }
 
+// setRequestGetBody sets GetBody function on the given http.Request if it is missing.  GetBody allows
+// reading from the request body without mutating the state of the request.
+func (a *lbAgent) setRequestBody(ctx context.Context, call *call) (*bytes.Buffer, error) {
+
+	r := call.req
+	if r.Body == nil || r.GetBody != nil {
+		return nil, nil
+	}
+
+	buf := bufPool.Get().(*bytes.Buffer)
+	buf.Reset()
+
+	// WARNING: we need to handle IO in a separate go-routine below
+	// to be able to detect a ctx timeout. When we timeout, we
+	// let gin/http-server to unblock the go-routine below.
+	errApp := make(chan error, 1)
+	go func() {
+
+		_, err := buf.ReadFrom(r.Body)
+		if err != nil {
+			errApp <- err
+			return
+		}
+
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(buf.Bytes()))
+
+		// GetBody does not mutate the state of the request body
+		r.GetBody = func() (io.ReadCloser, error) {
+			return ioutil.NopCloser(bytes.NewBuffer(buf.Bytes())), nil
```

Comment:
I'm still unclear about how this is safe.  Are we getting away with this because the NopCloser allows only reading from the buffer?

Reply:
Switching this to bytes.Reader() which should *really* be read-only.